docs/source/schemas/dynamic_columns.rst

```diff
@@ -0,0 +1,140 @@
+
+.. code:: java
+
+    TodoTable todoTable = TodoSchemaTableFactory.of().getTodoTable(tx);
+
+    todoTable.put(
+            TodoTable.TodoRow.of(person),
+            TodoTable.TodoColumnValue.of(
+                    TodoTable.TodoColumn.of(size1, cost1), description1),
+                    TodoTable.TodoColumn.of(size2, cost2), description2));
+
+Note that if the table already consists of a value for a given row-key and dynamic column values, then this put
+will overwrite the existing value.
+
+Retrieval
+=========
+
+One can use the standard ``getRowsMultimap`` method to retrieve a multimap of rows to column values. However, note
+that this method will call the underlying transaction's ``getRows`` method, which eagerly loads the entire row into
+memory!
+
+Alternatively, one can use ``getRowsColumnRange``. This takes a collection of rows and key ranges, and returns a map of
+row-keys to BatchingVisitables of column values. These may in turn be traversed using the BatchingVisitable
+API. For example, the code below is an implementation of query 1 for an arbitrary String ``person``:
+
+.. code:: java
+
+    TodoTable todoTable = TodoSchemaTableFactory.of().getTodoTable(tx);
+    Map<TodoTable.TodoRow, BatchingVisitable<TodoTable.TodoColumnValue>> results =
+            todoTable.getRowsColumnRange(
+                    ImmutableList.of(TodoTable.TodoRow.of(person)),
+                    BatchColumnRangeSelection.create(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY, 1));
+
+    if (!results.isEmpty()) {
+        AtomicReference<Todo> element = new AtomicReference<>();
+        BatchingVisitable<TodoTable.TodoColumnValue> visitable = Iterables.getOnlyElement(results.values());
+        visitable.batchAccept(1, item -> {
+            element.set(ImmutableTodo.of(item.get(0).getValue()));
+            return false; // Do not want any more, since we know our first batch contains the one.
+        });
+        return element.get();
+    }
+    return null;
+
+
+Data Representation (Cassandra)
+-------------------------------
+
+A row with dynamic columns is stored as a single Cassandra-level row. Recall the Atlas schema:
+
+.. code:: java
+
+    CREATE TABLE atlasete.default__todo (
+        key blob,
+        column1 blob,
+        column2 bigint,
+        value blob,
+        PRIMARY KEY (key, column1, column2)
+    )
+
+``key`` consists of the (encoded) row key; ``column1`` refers to the (encoded) column components, and ``column2``
+refers to the Atlas timestamp at which a write occurred. The clustering order of the table follows the
+``ValueByteOrder`` of the relevant column components.
+
+Conversely, observe that in a conventional (fixed columns) AtlasDB table, ``column1`` identifies which column is being
+encoded in the ``value``.
+
+When Not To Use Dynamic Columns
+-------------------------------
+
+Dynamic columns create wide rows in Cassandra, because they generate many (Cassandra) dynamic columns.
```

Comment:
It would be cool to have a recommendation for max row size here

Reply:
I think the technical limit is something like 2Bn, though would defer to C* experts (@tpetracca / @ashrayjain). As far as I'm aware the 'best practices' for this are something like <1M?

Reply:
(I've added a brief guideline; row size shouldn't exceed 1,000,000 dynamic column keys where possible.)

Reply:
from what i've heard and read, it's more about keeping the total size of the row < 100MB (this is also what i'm recommending in the "range scan best practices" doc)

Reply:
Yep. I should probably say more precisely what I've done: https://github.com/palantir/atlasdb/pull/2383/commits/150d8751a0b6d093b8df68e89ee50bdcb4c93349#diff-2db74fbb85be14b4da3a5278c51035beR150

(I've also heard that value as a recommendation being thrown around)

Reply:
I would say the limit is order 10's of MBs. We have a few internal users where it grows beyond that but their read patterns are constrained in a way that it doesn't cause the usual issues.